### PR TITLE
Exercise part 1

### PR DESCRIPTION
**Pull request by Aldaer**
artem_lodygin@epam.com

### CODE REVIEW

File: src/test/java/part1/exercise/CompletableFutureBasics.java [hunk 1]
```diff
@@ -60,42 +60,41 @@ public static void after() {
...
+        final Stream<Person> streamPerson = Stream.empty();
 
         final List<Person> persons = streamPerson.collect(toList());
         assertThat(persons.size(), is(0));
 
-        // TODO Complete CompletableFuture with NoSuchElementException
-        final CompletableFuture<Person> futurePerson = null;
-        // futurePerson.???
+        // Complete CompletableFuture with NoSuchElementException
+        final CompletableFuture<Person> futurePerson = CompletableFuture.completedFuture(optPerson).thenApply(Optional::get);
```
Comment (senia-psm) [inline src/test/java/part1/exercise/CompletableFutureBasics.java]: Use `CompletableFuture::completeExceptionally`

Comment (Aldaer) [reply src/test/java/part1/exercise/CompletableFutureBasics.java]: Fixed

File: src/main/java/part2/cache/CachingDataStorageImpl.java [hunk 2]
```diff
@@ -32,12 +31,23 @@ public CachingDataStorageImpl(DataStorage<String, T> db, int timeout, TimeUnit t
...
-        // TODO use remove(Object key, Object value) to remove target value
-        // TODO Start timeout after receiving result in CompletableFuture, not after receiving CompletableFuture itself
-        throw new UnsupportedOperationException();
+        final CompletableFuture<Void> outdated = new CompletableFuture<>();
+        final CompletableFuture<T> dbResponse = new CompletableFuture<>();
+        final OutdatableResult<T> result = new OutdatableResult<>(dbResponse, outdated);
+        final OutdatableResult<T> cachedResult = cache.putIfAbsent(key, result);
+        if (cachedResult != null) return cachedResult;
+
+        db.get(key).thenAccept(value -> {
```
Comment (senia-psm) [inline src/main/java/part2/cache/CachingDataStorageImpl.java:40]: Use `whenComplete` instead.

File: src/main/java/part2/cache/TypedEmployeeCachedStorage.java [hunk 3]
```diff
@@ -18,9 +26,93 @@ public TypedEmployeeCachedStorage(CachingDataStorage<String, data.Employee> empl
         this.employerStorage = employerStorage;
     }
 
+    private static class TypedEmployeeBuilder {
```
Comment (senia-psm) [inline src/main/java/part2/cache/TypedEmployeeCachedStorage.java:29]: Use `allOf` `anyOf` instead.

**Pull request closed without merging**